### PR TITLE
rescale! and center! along obsdim

### PR DESCRIPTION
**Pull request by abieler**
It might be useful to also have the feature rescaling functions working on both dimensions.
Currently the tests are failing, but I figured to check with you first if you like the general idea.

### CODE REVIEW

**Comment by Evizero** [comment]
Hi! Really cool that you are doing this. I agree that we should allow for the obsdim to be specified. In fact we already have a "system" for doing so which we use at [MLDataPatterns.jl](https://github.com/JuliaML/MLDataPattern.jl) (that package will be the new back-end for MLDataUtils for all data subsetting, k-folds etc).

Would be cool if you could adapt the code to that "system". I describe the general way of doing it here: https://github.com/joshday/OnlineStats.jl/issues/40#issuecomment-290209593 . For most code we allow any order array, but it would already be a big improvement to just have code for vectors an matrices

edit: `ObsDim` is defined in LearnBase.jl here: https://github.com/JuliaML/LearnBase.jl/blob/master/src/LearnBase.jl#L318-L395

**Comment by abieler** [comment]
Cool. I ll definitely try to adapt to that scheme!

**Comment by abieler** [comment]
somewhat like this?

File: src/feature_scaling.jl [hunk 1]
```diff
@@ -1,55 +1,130 @@
 """
-`μ = center!(X[, μ])`
+`μ = center!(X, obsdim[, μ])`
```
Comment (Evizero) [inline src/feature_scaling.jl]: `obsdim` should be depicted as the last parameter and also optional

File: src/feature_scaling.jl [hunk 2]
```diff
@@ -1,55 +1,130 @@
...
+function center!{T,N}(X::AbstractArray{T,N}, ::ObsDim.Last)
+  center!(X, ObsDim.Constant{N}())
+end
+
+function center!{T,N,M}(X::AbstractArray{T,N}, obsdim::ObsDim.Constant{M})
+  mu = vec(mean(X, M))
+  center!(X, mu, obsdim)
+end
+
+function center!{T}(X::AbstractVector{T}, bosdim::ObsDim.Constant{1})
```
Comment (Evizero) [inline src/feature_scaling.jl]: typo in `bosdim`, although you could just write `, ::ObsDim.Constant{1}`

File: src/feature_scaling.jl [hunk 3]
```diff
@@ -1,55 +1,130 @@
...
 If `μ` is not specified then it defaults to `mean(X, 2)`.
 """
-function center!{T<:AbstractFloat}(X::AbstractMatrix{T}, μ::AbstractVector = vec(mean(X, 2)))
-    k, n = size(X)
-    @inbounds for j = 1:k
-        for i = 1:n
-            X[j, i] = X[j, i] - μ[j]
-        end
+function center!(X, mu; obsdim=LearnBase.default_obsdim(X))
+  center!(X, mu, LearnBase.obs_dim(obsdim))
```
Comment (Evizero) [inline src/feature_scaling.jl]: the current default indentation for Julia is 4 spaces. so each indentation level should be 4 spaces more

File: src/feature_scaling.jl [hunk 4]
```diff
@@ -1,55 +1,130 @@
...
+  @inbounds for i in 1:length(X)
+    X[i] = X[i] - mu[i]
+  end
+  mu
+end
+   
+
 """
-`μ, σ = rescale!(X[, μ, σ])`
+`μ, σ = rescale!(X, obsdim[, μ, σ])`
```
Comment (Evizero) [inline src/feature_scaling.jl]: `obsdim` should be listed as the last parameter and also as optional

**Review by Evizero — CHANGES_REQUESTED** [review]
Looking good! Thanks for this. Made a few comments. Other than that the main thing missing are a few tests

**Comment by abieler** [comment]
Thanks for the comments, I ll add some tests later on

File: src/feature_scaling.jl [hunk 5]
```diff
@@ -1,57 +1,146 @@
...
+Centers `X` along obsdim around the corresponding entry in the vector `μ`.
 If `μ` is not specified then it defaults to `mean(X, 2)`.
 """
-function center!{T<:AbstractFloat}(X::AbstractMatrix{T}, μ::AbstractVector = vec(mean(X, 2)))
-    k, n = size(X)
-    @inbounds for j = 1:k
-        for i = 1:n
-            X[j, i] = X[j, i] - μ[j]
+function center!(X, μ; obsdim=LearnBase.default_obsdim(X))
+    center!(X, μ, LearnBase.obs_dim(obsdim))
```
Comment (Evizero) [inline src/feature_scaling.jl]: can you change 

```julia
LearnBase.obs_dim(obsdim)
```

everywhere to

```julia
convert(LearnBase.ObsDimension, obsdim)
```

(I deprecated the former just a few days ago)

**Review by Evizero — CHANGES_REQUESTED** [review]
lgtm! Nice work. One last request

File: src/feature_scaling.jl [hunk 6]
```diff
@@ -127,14 +127,14 @@ function rescale!(X::AbstractMatrix, μ::AbstractVector, σ::AbstractVector, ::O
     μ, σ
 end
 
-function rescale!(X::AbstractVector, μ::AbstractVector, σ::AbstractVector, ::ObsDim.Constant{1})
```
Comment (Evizero) [inline src/feature_scaling.jl]: why did you remove this?

Comment (abieler) [reply src/feature_scaling.jl]: i figured this would include ObsDim.First() and .Last(). No?

Comment (Evizero) [reply src/feature_scaling.jl]: Well, since `X` is a vector, the fallback method you implemented transforms `ObsDim.Last` to `ObsDim.Constant{1}` anyway.

I think here `::ObsDim.Constant{1}` was cleaner than `::ObsDim.Constant` (which would also allow `ObsDim.Constant{2}()`), because if the data `X` is a vector, then `1` is the only sensible dimension to choose. i.e. I'd expect a method error if I use `ObsDim.Constant{2}()`.

Comment (abieler) [reply src/feature_scaling.jl]: makes sense. i ll undo my improvements :)

**Comment by Evizero** [comment]
This is optional, but if you feel up to it, it would be cool to update the corresponding section in the documentation: https://raw.githubusercontent.com/JuliaML/MLDataUtils.jl/master/docs/data/feature.rst

**Review by Evizero — APPROVED** [review]

**Comment by Evizero** [comment]
lgtm. Ready to merge when you are

**Comment by Evizero** [comment]
thanks!

**Comment by abieler** [comment]
thanks for all the comments! also learned about singleton types in the process. :)
how do you feel about support for dataframes and datatables? worth looking at or do you want to keep this for arrays only?

**Comment by Evizero** [comment]
> how do you feel about support for dataframes and datatables

I will add a `DataFrames` dependency in the next update (see `dev` branch), so I am open to the idea.
